src/common/base32.cpp

```diff
@@ -0,0 +1,506 @@
+// All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without modification, are
+// permitted provided that the following conditions are met:
+//
+// 1. Redistributions of source code must retain the above copyright notice, this list of
+//    conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright notice, this list
+//    of conditions and the following disclaimer in the documentation and/or other
+//    materials provided with the distribution.
+//
+// 3. Neither the name of the copyright holder nor the names of its contributors may be
+//    used to endorse or promote products derived from this software without specific
+//    prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+// MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+// THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+// STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+// THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+/**
+ *  Copyright (C) 2015 Trustifier Inc.
+ *  Copyright (C) 2015 Ahmed Masud
+ *  Copyright (C) 2015 Topology LP
+ *  Copyright (C) 2018 Jakob Petsovits
+ *  All rights reserved.
+ *
+ *  Permission is hereby granted, free of charge, to any person obtaining a copy
+ *  of this software and associated documentation files (the "Software"), to
+ *  deal in the Software without restriction, including without limitation the
+ *  rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
+ *  sell copies of the Software, and to permit persons to whom the Software is
+ *  furnished to do so, subject to the following conditions:
+ *
+ *  The above copyright notice and this permission notice shall be included in
+ *  all copies or substantial portions of the Software.
+ *
+ *  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ *  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ *  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL
+ *  THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ *  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ *  FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+ *  IN THE SOFTWARE.
+ *
+ *  Adapted from https://github.com/ahmed-masud/libbase32,
+ *  commit 79761b2b79b0545697945efe0987a8d3004512f9.
+ *  Quite different now.
```

Comment:
> Quite different now.

I wonder quite a bit about this. Why, explained on a high conceptual level, did the code have to change much? The different alphabet is clear of course, but beyond that?

I have a hard time anyway to compare your code with the forked original code, which makes it very hard to quickly decide whether the code is still correct without basically looking at everything. The original ist this file, right? https://github.com/ahmed-masud/libbase32/blob/master/base32.c

Reply:
Ok, misunderstand from my side, this line is not from you, @DangerousFreedom1984 , but already present in the code that you forked:

> Adapted from https://github.com/ahmed-masud/libbase32

As far as I see you don't give info in this PR where you got the code from, unfortunately, but I think I found it here: https://github.com/tplgy/cppcodec/blob/master/cppcodec/detail/base32.hpp , with parts from other files of that repo copied in, to have it all in one file I suppose.

Which still leads me to the question: That file from Ahmed Masud I linked above is wonderfully small, and wonderfully readable. Is it really not fit for purpose?

Reply:
Yes, I copied most of the code from https://github.com/tplgy/cppcodec/blob/master/cppcodec/detail/base32.hpp as it seems to me the most reliable and efficient base32 library.
The code from Ahmed does not seem to be as efficient as this one as they improved it. I may be wrong though.
Since both are MIT licenses I believe that the correct way to copy is by pasting the exact header in our header file, right?

Reply:
> Since both are MIT licenses I believe that the correct way to copy is by pasting the exact header in our header file, right?

I am a unsure myself. The wording of the license is a bit different, maybe just a different iteration. At least we are on the safe side if we copy verbatim I guess.